theseus/labs/lie/lie_tensor.py

```diff
@@ -0,0 +1,272 @@
+    #   - `lie.lift(ltype, matrix)`
+    # that we expect to be more commonly used.
+    def exp(self, tangent_vector: torch.Tensor) -> "LieTensor":
+        return self.new(_eval_op(self._fn_lib, "exp", tangent_vector))
+
+    def hat(self, tangent_vector: torch.Tensor) -> torch.Tensor:
+        return _eval_op(self._fn_lib, "hat", tangent_vector)
+
+    def vee(self, matrix: torch.Tensor) -> torch.Tensor:
+        return _eval_op(self._fn_lib, "vee", matrix)
+
+    def lift(self, matrix: torch.Tensor) -> torch.Tensor:
+        return _eval_op(self._fn_lib, "lift", matrix)
+
+    def project(self, matrix: torch.Tensor) -> torch.Tensor:
+        return _eval_op(self._fn_lib, "project", matrix)
+
+    # For the next ones the output also depends on self's data
+    def log(self) -> torch.Tensor:
+        return self._fn_lib.log(self._t)
+
+    def adj(self) -> torch.Tensor:
+        return self._fn_lib.adj(self._t)
+
+    def inv(self) -> "LieTensor":
+        return self.new(self._fn_lib.inv(self._t))
+
+    def compose(self, other: "LieTensor") -> "LieTensor":
+        self._check_ltype(other, "compose")
+        return self.new(self._fn_lib.compose(self._t, other._t))
+
+    def left_act(self, matrix: torch.Tensor) -> torch.Tensor:
+        return self._fn_lib.left_act(self._t, matrix)
+
+    def left_project(self, matrix: torch.Tensor) -> torch.Tensor:
+        return self._fn_lib.left_project(self._t, matrix)
+
+    # Operator Jacobians
+    def _unary_jop_base(
+        self,
+        input0: torch.Tensor,
+        fn: UnaryOperatorOpFnType,
+        out_is_group: bool = True,
+    ) -> _JFnReturnType:
+        jacs: List[torch.Tensor] = []
+        op_res: TensorType = fn(input0, jacobians=jacs)
+        if out_is_group:
+            op_res = self.new(op_res)
+        return jacs, op_res
+
+    def jexp(self, tangent_vector: torch.Tensor) -> _JFnReturnType:
+        return self._unary_jop_base(tangent_vector, self._fn_lib.exp)
+
+    def jlog(self) -> _JFnReturnType:
+        return self._unary_jop_base(self._t, self._fn_lib.log, out_is_group=False)
+
+    def jinv(self) -> _JFnReturnType:
+        return self._unary_jop_base(self._t, self._fn_lib.inv)
+
+    def jcompose(self, other: "LieTensor") -> _JFnReturnType:
+        self._check_ltype(other, "jcompose")
+        jacs: List[torch.Tensor] = []
+        op_res = self.new(self._fn_lib.compose(self._t, other._t, jacobians=jacs))
+        return jacs, op_res
+
+    def _no_jop(self, input0: TensorType) -> _JFnReturnType:
+        raise NotImplementedError
+
+    jadjoint = _no_jop
+    jhat = _no_jop
+    jvee = _no_jop
+    jlift = _no_jop
+    jproject = _no_jop
+    jleft_act = _no_jop
+    jleft_project = _no_jop
+
+
+# ----------------------------
+# Tensor creation functions
+# ----------------------------
+def new(
+    data: Any,
+    ltype: Optional[_ltype] = None,
+    dtype: torch.dtype = torch.float,
+    device: Union[torch.device, str] = None,
```

Comment:
Can dtype and device be optional since they have sensible defaults in the function signature? These could also possibly conflict with what a passed torch tensor might have.

Reply:
Not sure I understand the question. If the question is about who has priority between the passed torch tensor and the given dtype/device, I think the priority should be the explicit values passed by the user (will need to add this, not currently the case).